File: macOS/ai/Data/ApiClient.swift
import Foundation
import os

/// URLSession-based API client replacing Retrofit + OkHttp.
/// Handles auth per format: Bearer (OpenAI), x-api-key (Claude), ?key= query param (Gemini).
actor ApiClient {
    static let shared = ApiClient()

    private static let logger = Logger(subsystem: "com.ai.macAI", category: "ApiClient")

    private let session: URLSession

    private init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 600  // 10 min for long AI responses
        self.session = URLSession(configuration: config)
    }

    // MARK: - Generic Request

    /// Send a JSON request and decode the response.
    func request<Req: Encodable, Res: Decodable>(
        url: String,
        method: String = "POST",
        body: Req,
        service: AppService,
        apiKey: String,
        extraHeaders: [String: String] = [:]
    ) async throws -> (Res, HTTPURLResponse) {
        guard let requestURL = URL(string: url) else {
            throw ApiError.invalidURL(url)
        }

        var request = URLRequest(url: requestURL)
        request.httpMethod = method

        // Set auth headers based on API format
        switch service.apiFormat {
        case .anthropic:
            request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
            request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        case .google:
            // Google uses ?key= query param, handled in URL construction
            break
        case .openaiCompatible:
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        }

        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        // Extra headers
        for (key, value) in extraHeaders {
            request.setValue(value, forHTTPHeaderField: key)
        }

        // Encode body
        let encoder = JSONEncoder()
        request.httpBody = try encoder.encode(body)

        // Trace request if enabled
        let traceTimestamp = Date()
        let requestBodyString = String(data: request.httpBody ?? Data(), encoding: .utf8)

        // Execute
        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw ApiError.invalidResponse
        }

        // Trace response if enabled
        await ApiTracer.shared.traceIfEnabled(
            url: url,
            method: method,
            requestHeaders: request.allHTTPHeaderFields ?? [:],
            requestBody: requestBodyString,
            statusCode: httpResponse.statusCode,
            responseHeaders: httpResponse.allHeaderFields as? [String: String] ?? [:],
            responseBody: String(data: data, encoding: .utf8),
            timestamp: traceTimestamp
        )

        // Decode response
        let decoder = JSONDecoder()
        do {
            let decoded = try decoder.decode(Res.self, from: data)
            return (decoded, httpResponse)
        } catch {
            // Try to get error message from raw response
            let bodyStr = String(data: data, encoding: .utf8) ?? "Unknown"
            Self.logger.error("Decode error for \(url): \(error.localizedDescription). Body: \(bodyStr.prefix(500))")
            throw ApiError.decodingError(error, bodyStr)
        }
    }

    // MARK: - GET Request

    func get<Res: Decodable>(
        url: String,
        service: AppService,
        apiKey: String
    ) async throws -> (Res, HTTPURLResponse) {
        guard var requestURL = URL(string: url) else {
            throw ApiError.invalidURL(url)
        }

        var request: URLRequest

        switch service.apiFormat {
        case .google:
            // Add ?key= query param for Gemini
            var components = URLComponents(url: requestURL, resolvingAgainstBaseURL: false)!
            var items = components.queryItems ?? []
            items.append(URLQueryItem(name: "key", value: apiKey))
            components.queryItems = items
            requestURL = components.url!
            request = URLRequest(url: requestURL)
        case .anthropic:
            request = URLRequest(url: requestURL)
            request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
            request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        case .openaiCompatible:
            request = URLRequest(url: requestURL)
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        }

        request.httpMethod = "GET"

        let (data, response) = try await session.data(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ApiError.invalidResponse
        }

        let decoded = try JSONDecoder().decode(Res.self, from: data)
        return (decoded, httpResponse)
    }

    // MARK: - Streaming Request

    /// Send a streaming request and return raw bytes for SSE parsing.
    func streamRequest<Req: Encodable>(
        url: String,
        body: Req,
        service: AppService,
        apiKey: String
    ) async throws -> (URLSession.AsyncBytes, HTTPURLResponse) {
        guard let requestURL = URL(string: url) else {
            throw ApiError.invalidURL(url)
        }

        var request = URLRequest(url: requestURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        switch service.apiFormat {
        case .anthropic:
            request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
            request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        case .google:
            break
        case .openaiCompatible:
            request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        }

        request.httpBody = try JSONEncoder().encode(body)

        let (bytes, response) = try await session.bytes(for: request)
        guard let httpResponse = response as? HTTPURLResponse else {
            throw ApiError.invalidResponse
        }

        return (bytes, httpResponse)
    }

    // MARK: - Build URLs

    /// Build a chat completions URL for a provider with optional custom base URL.
    static func chatUrl(for service: AppService, customBaseUrl: String? = nil) -> String {
        let base = customBaseUrl ?? service.baseUrl
        let normalizedBase = base.hasSuffix("/") ? base : "\(base)/"
        return normalizedBase + service.chatPath
    }

    /// Build a Gemini URL for generate content.
    static func geminiGenerateUrl(baseUrl: String, model: String, apiKey: String, stream: Bool = false) -> String {
        let base = baseUrl.hasSuffix("/") ? baseUrl : "\(baseUrl)/"
        let action = stream ? "streamGenerateContent" : "generateContent"
        var url = "\(base)v1beta/models/\(model):\(action)?key=\(apiKey)"
        if stream { url += "&alt=sse" }
        return url
    }

    /// Build a models list URL for a provider.
    static func modelsUrl(for service: AppService, customUrl: String? = nil) -> String {
        if let custom = customUrl, !custom.isEmpty { return custom }
        guard let modelsPath = service.modelsPath else { return "" }
        let base = service.baseUrl.hasSuffix("/") ? service.baseUrl : "\(service.baseUrl)/"
        return base + modelsPath
    }
}

// MARK: - API Errors

enum ApiError: LocalizedError {
    case invalidURL(String)
    case invalidResponse
    case httpError(Int, String)
    case decodingError(Error, String)
    case noApiKey
    case streamError(String)

    var errorDescription: String? {
        switch self {
        case .invalidURL(let url): return "Invalid URL: \(url)"
        case .invalidResponse: return "Invalid response from server"
        case .httpError(let code, let msg): return "HTTP \(code): \(msg)"
        case .decodingError(let error, let body): return "Decoding error: \(error.localizedDescription). Body: \(body.prefix(200))"
        case .noApiKey: return "API key not configured"
        case .streamError(let msg): return "Stream error: \(msg)"
        }
    }
}

File: windows/AI/Data/ApiClient.cs
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace AI.Data;

/// <summary>
/// HttpClient-based API client replacing URLSession.
/// Handles auth per format: Bearer (OpenAI), x-api-key + anthropic-version (Anthropic),
/// ?key= query param (Google).
/// </summary>
public class ApiClient
{
    public static readonly ApiClient Instance = new();

    private static readonly HttpClient Http;

    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = null,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
    };

    static ApiClient()
    {
        Http = new HttpClient
        {
            Timeout = TimeSpan.FromSeconds(600)  // Resource timeout; per-request timeout via CancellationToken
        };
    }

    private ApiClient() { }

    // MARK: - Generic POST Request

    /// <summary>
    /// Send a JSON POST request, decode the response, and trace via ApiTracer.
    /// </summary>
    public async Task<(TRes Response, int StatusCode)> RequestAsync<TReq, TRes>(
        string url,
        TReq body,
        AppService service,
        string apiKey,
        Dictionary<string, string>? extraHeaders = null)
    {
        var requestUrl = BuildUrl(url, service, apiKey);

        using var request = new HttpRequestMessage(HttpMethod.Post, requestUrl);
        ApplyAuthHeaders(request, service, apiKey);
        request.Content = new StringContent(
            JsonSerializer.Serialize(body, JsonOptions),
            Encoding.UTF8,
            "application/json");

        if (extraHeaders != null)
        {
            foreach (var (key, value) in extraHeaders)
                request.Headers.TryAddWithoutValidation(key, value);
        }

        var requestBodyString = await (request.Content?.ReadAsStringAsync() ?? Task.FromResult(""));
        var traceTimestamp = DateTime.UtcNow;

        using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30));
        HttpResponseMessage response;
        try
        {
            response = await Http.SendAsync(request, HttpCompletionOption.ResponseContentRead, cts.Token);
        }
        catch (OperationCanceledException)
        {
            throw new ApiError(ApiErrorKind.HttpError, "Request timed out (30s)");
        }

        var statusCode = (int)response.StatusCode;
        var responseBody = await response.Content.ReadAsStringAsync();

        // Collect headers for tracing
        var requestHeaders = CollectHeaders(request.Headers);
        var responseHeaders = CollectHeaders(response.Headers);

        await ApiTracer.Instance.TraceIfEnabledAsync(
            url: requestUrl,
            method: "POST",
            requestHeaders: requestHeaders,
            requestBody: requestBodyString,
            statusCode: statusCode,
            responseHeaders: responseHeaders,
            responseBody: responseBody,
            timestamp: traceTimestamp);

        if (!response.IsSuccessStatusCode)
        {
            throw new ApiError(ApiErrorKind.HttpError, $"HTTP {statusCode}: {responseBody}");
        }

        try
        {
            var decoded = JsonSerializer.Deserialize<TRes>(responseBody, JsonOptions)
                ?? throw new ApiError(ApiErrorKind.DecodingError, "Deserialized to null");
            return (decoded, statusCode);
        }
        catch (JsonException ex)
        {
            var preview = responseBody.Length > 500 ? responseBody[..500] : responseBody;
            throw new ApiError(ApiErrorKind.DecodingError, $"Deserialization error: {ex.Message}. Body: {preview}");
        }
    }

    // MARK: - GET Request

    /// <summary>
    /// Send a GET request with format-specific auth and decode the response.
    /// </summary>
    public async Task<(TRes Response, int StatusCode)> GetAsync<TRes>(
        string url,
        AppService service,
        string apiKey)
    {
        var requestUrl = BuildUrl(url, service, apiKey);

        using var request = new HttpRequestMessage(HttpMethod.Get, requestUrl);
        ApplyAuthHeaders(request, service, apiKey);

        using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30));
        HttpResponseMessage response;
        try
        {
            response = await Http.SendAsync(request, HttpCompletionOption.ResponseContentRead, cts.Token);
        }
        catch (OperationCanceledException)
        {
            throw new ApiError(ApiErrorKind.HttpError, "Request timed out (30s)");
        }

        var statusCode = (int)response.StatusCode;
        if (!response.IsSuccessStatusCode)
        {
            var errorBody = await response.Content.ReadAsStringAsync();
            throw new ApiError(ApiErrorKind.HttpError, $"HTTP {statusCode}: {errorBody}");
        }

        var responseBody = await response.Content.ReadAsStringAsync();

        try
        {
            var decoded = JsonSerializer.Deserialize<TRes>(responseBody, JsonOptions)
                ?? throw new ApiError(ApiErrorKind.DecodingError, "Deserialized to null");
            return (decoded, statusCode);
        }
        catch (JsonException ex)
        {
            var preview = responseBody.Length > 500 ? responseBody[..500] : responseBody;
            throw new ApiError(ApiErrorKind.DecodingError, $"Deserialization error: {ex.Message}. Body: {preview}");
        }
    }

    // MARK: - Streaming Request

    /// <summary>
    /// Send a POST request and return the HttpResponseMessage for SSE streaming.
    /// Caller is responsible for disposing the response.
    /// </summary>
    public async Task<HttpResponseMessage> StreamRequestAsync<TReq>(
        string url,
        TReq body,
        AppService service,
        string apiKey)
    {
        var requestUrl = BuildUrl(url, service, apiKey);

        var request = new HttpRequestMessage(HttpMethod.Post, requestUrl);
        ApplyAuthHeaders(request, service, apiKey);
        request.Content = new StringContent(
            JsonSerializer.Serialize(body, JsonOptions),
            Encoding.UTF8,
            "application/json");

        HttpResponseMessage response;
        try
        {
            // ResponseHeadersRead: return as soon as headers arrive; body streams from caller
            response = await Http.SendAsync(request, HttpCompletionOption.ResponseHeadersRead);
        }
        catch (OperationCanceledException)
        {
            throw new ApiError(ApiErrorKind.StreamError, "Stream request timed out");
        }

        if (!response.IsSuccessStatusCode)
        {
            var errorBody = await response.Content.ReadAsStringAsync();
            response.Dispose();
            throw new ApiError(ApiErrorKind.HttpError, $"HTTP {(int)response.StatusCode}: {errorBody}");
        }

        return response;  // Caller owns disposal
    }

    // MARK: - Auth Helpers

    private static void ApplyAuthHeaders(HttpRequestMessage request, AppService service, string apiKey)
    {
        switch (service.ApiFormat)
        {
            case ApiFormat.Anthropic:
                request.Headers.TryAddWithoutValidation("x-api-key", apiKey);
                request.Headers.TryAddWithoutValidation("anthropic-version", "2023-06-01");
                break;
            case ApiFormat.Google:
                // Auth via ?key= query param appended in BuildUrl; no header needed
                break;
            case ApiFormat.OpenAiCompatible:
            default:
                request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);
                break;
        }
    }

    /// <summary>
    /// For Google, appends ?key= to the URL. For all other formats returns url unchanged.
    /// </summary>
    private static string BuildUrl(string url, AppService service, string apiKey)
    {
        if (service.ApiFormat != ApiFormat.Google) return url;

        var separator = url.Contains('?') ? "&" : "?";
        return $"{url}{separator}key={Uri.EscapeDataString(apiKey)}";
    }

    private static Dictionary<string, string> CollectHeaders(HttpHeaders headers)
    {
        var result = new Dictionary<string, string>();
        foreach (var (key, values) in headers)
            result[key] = string.Join(", ", values);
        return result;
    }

    // MARK: - URL Builders

    /// <summary>
    /// Build a chat completions URL for a provider with optional custom base URL.
    /// </summary>
    public static string ChatUrl(AppService service, string? customBaseUrl = null)
    {
        var baseUrl = customBaseUrl ?? service.BaseUrl;
        var normalizedBase = baseUrl.EndsWith('/') ? baseUrl : baseUrl + "/";
        return normalizedBase + service.ChatPath;
    }

    /// <summary>
    /// Build a Gemini generateContent or streamGenerateContent URL.
    /// The ?key= param is embedded here for Gemini since the URL is constructed
    /// externally before being passed to the streaming/request methods.
    /// </summary>
    public static string GeminiGenerateUrl(string baseUrl, string model, string apiKey, bool stream = false)
    {
        var normalizedBase = baseUrl.EndsWith('/') ? baseUrl : baseUrl + "/";
        var action = stream ? "streamGenerateContent" : "generateContent";
        var url = $"{normalizedBase}v1beta/models/{model}:{action}?key={Uri.EscapeDataString(apiKey)}";
        if (stream) url += "&alt=sse";
        return url;
    }

    /// <summary>
    /// Build a models list URL for a provider, with optional custom URL override.
    /// </summary>
    public static string ModelsUrl(AppService service, string? customUrl = null)
    {
        if (!string.IsNullOrEmpty(customUrl)) return customUrl;
        if (string.IsNullOrEmpty(service.ModelsPath)) return "";
        var normalizedBase = service.BaseUrl.EndsWith('/') ? service.BaseUrl : service.BaseUrl + "/";
        return normalizedBase + service.ModelsPath;
    }
}

// MARK: - API Errors

public enum ApiErrorKind
{
    InvalidUrl,
    InvalidResponse,
    HttpError,
    DecodingError,
    NoApiKey,
    StreamError
}

public class ApiError : Exception
{
    public ApiErrorKind Kind { get; }

    public ApiError(ApiErrorKind kind, string detail = "") : base(BuildMessage(kind, detail))
    {
        Kind = kind;
    }

    private static string BuildMessage(ApiErrorKind kind, string detail) => kind switch
    {
        ApiErrorKind.InvalidUrl      => $"Invalid URL: {detail}",
        ApiErrorKind.InvalidResponse => "Invalid response from server",
        ApiErrorKind.HttpError       => detail,   // already formatted as "HTTP NNN: ..."
        ApiErrorKind.DecodingError   => detail,   // already formatted with context
        ApiErrorKind.NoApiKey        => "API key not configured",
        ApiErrorKind.StreamError     => $"Stream error: {detail}",
        _                            => detail
    };
}
